app/assets/javascripts/admin/addon/components/emoji-value-list.js

```diff
@@ -0,0 +1,198 @@
+  values: null,
+  validationMessage: null,
+  emojiPickerIsActive: false,
+  isEditorFocused: false,
+  emojiName: null,
+
+  init() {
+    this._super(...arguments);
+    this.set("collection", []);
+  },
+
+  @action
+  closeEmojiPicker() {
+    this.collection.setEach("isEditing", false);
+    this.set("emojiPickerIsActive", false);
+    this.set("isEditorFocused", false);
+  },
+
+  @action
+  emojiSelected(code) {
+    if (!this._validateInput(code)) {
+      return;
+    }
+
+    const item = this.collection.findBy("isEditing");
+    if (item) {
+      setProperties(item, {
+        value: code,
+        emojiUrl: emojiUrlFor(code),
+        isEditing: false,
+      });
+
+      this._saveValues();
+    } else {
+      this.set("emojiName", code);
+    }
+
+    this.set("emojiPickerIsActive", false);
+    this.set("isEditorFocused", false);
+  },
+
+  @action
+  clearInput() {
+    this.set("emojiName", null);
+  },
+
+  @on("didReceiveAttrs")
+  _setupCollection() {
+    this.set("collection", this._splitValues(this.values));
+  },
+
+  @discourseComputed("collection")
+  showUpDownButtons(collection) {
+    return collection.length - 1 ? true : false;
+  },
+
+  _splitValues(values) {
+    if (values && values.length) {
+      const emojiList = [];
+      const emojis = values.split("|").filter(Boolean);
+      emojis.forEach((emojiName) => {
+        const emoji = {
+          isEditable: true,
+          isEditing: false,
+        };
+        emoji.value = emojiName;
+        emoji.emojiUrl = emojiUrlFor(emojiName);
+
+        emojiList.push(emoji);
+      });
+
+      return emojiList;
+    } else {
+      return [];
+    }
+  },
+
+  @action
+  editValue(index) {
+    this.closeEmojiPicker();
+    schedule("afterRender", () => {
+      if (parseInt(index, 10) >= 0) {
+        const item = this.collection[index];
+        if (item.isEditable) {
+          set(item, "isEditing", true);
+        }
+      }
+
+      this.set("isEditorFocused", true);
+      later(() => {
+        if (!this.element || !this.isDestroyed || !this.isDestroying) {
+          this.set("emojiPickerIsActive", true);
+        }
+      }, 100);
+    });
+  },
+
+  @action
+  addValue() {
+    if (!this._validateInput(this.emojiName)) {
+      return;
+    }
+
+    this._addValue(this.emojiName);
+    this.set("emojiName", null);
+  },
+
+  @action
+  removeValue(value) {
+    this._removeValue(value);
+  },
+
+  rotateCollection(direction) {
+    if (direction < 0) {
+      this.collection.push(this.collection.shift());
+    } else {
+      this.collection.unshift(this.collection.pop());
+    }
+
+    this._saveValues();
+  },
+
+  @action
+  shift(operation, index) {
+    if (!operation) {
+      return;
+    }
+
+    // rotate if shift is up and index is 0
+    if (!index && operation < 0) {
+      return this.rotateCollection(-1);
+    }
+
+    // rotate if shift is down and index is last
+    if (index === this.collection.length - 1 && operation > 0) {
```

Comment:
Knowing this:

![Capture d’écran 2021-04-06 à 08 56 57](https://user-images.githubusercontent.com/339945/113671057-28f08e00-96b6-11eb-8a68-b776e2ce7ad0.png)

You will see that you can simplify this code